### PR TITLE
Implement sensitivities for `BLAS.gemm`

### PR DESCRIPTION
**Pull request by ararslan**
These are ported from Nabla.

### CODE REVIEW

File: src/rules/blas.jl [hunk 1]
```diff
@@ -72,3 +74,37 @@ function rrule(f::typeof(BLAS.gemv), tA, A, x)
...
 end
+
+#####
+##### `BLAS.gemm`
+#####
+
+function rrule(::typeof(gemm), tA::Char, tB::Char, α::Number,
+               A::AbstractMatrix, B::AbstractMatrix)
+    C = gemm(tA, tB, α, A, B)
+    #∂α = C̄ -> sum(C̄ .* C) / α  TODO: This is wrong and idk why
```
Comment (willtebbutt) [inline src/rules/blas.jl]: Should this be
```julia
∂α = C̄ -> sum(C̄ .* C)
```

Comment (ararslan) [reply src/rules/blas.jl]: Unfortunately that doesn't seem to do it:
```
gemm: Test Failed at /home/alex/.julia/dev/ChainRules/test/rules/blas.jl:22
  Expression: ≈(x̄_ad, x̄_fd, rtol=1.0e-9, atol=1.0e-9)
   Evaluated: 4.6207100689748355 ≈ -1.8353251336788157 (rtol=1.0e-9, atol=1.0e-9)
```

Comment (ararslan) [reply src/rules/blas.jl]: Dividing by `α` is what Nabla does; I have no idea why it doesn't work in this instance.

Comment (willtebbutt) [reply src/rules/blas.jl]: Hmm yeah, this is really weird. Frankly, I have no idea why Nabla does it...

**Comment by willtebbutt** [comment]
I've also just realised that the typing is probably a bit loose. I think that BLAS routines are usually restricted to strided arrays with particular types of floating-point elements. Perhaps we should do the same here?

File: src/rules/blas.jl (hunk 1 above)
Comment (ararslan) [reply src/rules/blas.jl]: Perhaps we should add a `NotImplementedRule` and punt it down the road :stuck_out_tongue:

**Comment by ararslan** [comment]
`gemm` is a case where the methods in LinearAlgebra are, for whatever reason, specified as `AbstractMatrix` instead of `StridedMatrix`. That is a good observation about the element type; it's restricted to `BlasFloat`, and with `alpha`'s type matching the element type of the matrices.

File: src/rules/blas.jl (hunk 1 above)
Comment (ararslan) [reply src/rules/blas.jl]: > Frankly, I have no idea why Nabla does it

Tbh I don't either. I don't see how it would arise from the math.

Comment (ararslan) [reply src/rules/blas.jl]: > Perhaps we should add a NotImplementedRule and punt it down the road 😛

I couldn't figure out how to deal with alpha so I actually did this in #28.

**Comment by ararslan** [comment]
To do:
* Add tests for non-square matrices

**Comment by ararslan** [comment]
Rectangular tests added. Good to go?

**Comment by willtebbutt** [comment]
LGTM

**Pull request merged**